internal/controllers/extension_controller.go

```diff
@@ -182,3 +426,89 @@ func extensionRequestsForCatalog(c client.Reader, logger logr.Logger) handler.Ma
 		return requests
 	}
 }
+
+func (r *ExtensionReconciler) GenerateExpectedApp(o ocv1alpha1.Extension, bundlePath string) *unstructured.Unstructured {
+	// We use unstructured here to avoid problems of serializing default values when sending patches to the apiserver.
+	// If you use a typed object, any default values from that struct get serialized into the JSON patch, which could
+	// cause unrelated fields to be patched back to the default value even though that isn't the intention. Using an
+	// unstructured ensures that the patch contains only what is specified. Using unstructured like this is basically
+	// identical to "kubectl apply -f"
+	spec := map[string]interface{}{
+		"serviceAccountName": o.Spec.ServiceAccountName,
+		"fetch": []interface{}{
+			map[string]interface{}{
+				"image": map[string]interface{}{
+					"url": bundlePath,
+				},
+			},
+		},
+		"template": []interface{}{
+			map[string]interface{}{
+				"ytt": map[string]interface{}{},
+			},
+		},
+		"deploy": []interface{}{
+			map[string]interface{}{
+				"kapp": map[string]interface{}{},
+			},
+		},
+	}
+
+	app := &unstructured.Unstructured{
+		Object: map[string]interface{}{
+			"apiVersion": "kappctrl.k14s.io/v1alpha1",
+			"kind":       "App",
+			"metadata": map[string]interface{}{
+				"name":      o.GetName(),
+				"namespace": o.GetNamespace(),
+			},
+			"spec": spec,
+		},
+	}
+
+	app.SetOwnerReferences([]metav1.OwnerReference{
+		{
+			APIVersion:         ocv1alpha1.GroupVersion.String(),
+			Kind:               "Extension",
+			Name:               o.Name,
+			UID:                o.UID,
+			Controller:         ptr.To(true),
+			BlockOwnerDeletion: ptr.To(true),
+		},
+	})
+	return app
+}
+
+func (r *ExtensionReconciler) variables(ctx context.Context) ([]deppy.Variable, error) {
+	allBundles, err := r.BundleProvider.Bundles(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	extensionList := ocv1alpha1.ExtensionList{}
+	if err := r.Client.List(ctx, &extensionList); err != nil {
+		return nil, err
+	}
+
+	// Filter apps belogining to current namespace only to create install variables?
+	// This would ignore cluster scoped APIs installed from other Apps.
```

Comment:
I think we're going to let kapp-controller/kapp logic enforce the "an object can only be owned by one extension rule" because kapp already has that rule for "an object can only be owned by one app"